Fix various background inconsistencies

barklund opened the pull request:
## Summary

Fix a number of inconsistencies reported about background setting and removing.

Note that this does not yet reflect the changes coming up after #1207 is merged - in particular for this PR overlay still belongs to and sticks with the page, not the individual background image.

## Relevant Technical Choices

For one, this makes sure to use `combineElements` when clicking "Set as background" to make sure same code path is followed.

Then make sure flip is also copied when combining elements, as the new element should not inherit the flip of the old element, but rather keep its own setting.

Finally, this fetches more info about currently dragged image when showing temp drop replacements. This required reworking some things in `DisplayElement` to make sure replacement mask is correctly flipped or not.

## To-do

<!-- A list of things that need to be addressed in this PR or follow-up changes. -->

## User-facing changes

This makes sure, that no matter how you set an image as background, you get the same consistent behavior on a number of topics - whatever that is decided to be.

## Testing Instructions


--

<!-- Please reference the issue(s) this PR addresses. -->

Fixes #1411
Fixes #1821

Commit:
Copy properties from original when temp replacing

The properties flip, scale, focusX and focusY will be copied from the original element when previewing. Because flip is handled by the mask, the replacement element has to be outside the mask and come with its own mask.

Finally this also uses `combineElements` when setting as background, to unify the way background is set whether you drag or use the button to do it.

Commit:
Update tests

barklund commented on assets/src/edit-story/components/canvas/displayElement.js:
```diff
@@ -119,14 +120,24 @@ function DisplayElement({ element, previewMode, page }) {
+            </WithMask>
+          )}
+        </ReplacementContainer>
+      )}
```
So there's  slight issue with this - when you hover over a blob with a different flip setting the dragging image has, the blob can be seen in both states at the same time. This can be fixed by wrapping the "real" display element in a container similar to `ReplacementContainer`, but it just fades "the other way".

@wassgha, WDYT here? I think you made this originally.

miina replied:
@barklund Note there's also an existing issue with flipped shapes: https://github.com/google/web-stories-wp/issues/1929

barklund replied:
I say we punt this to that one. Adding a mention there as well.

barklund replied:
Or actually in #1111.

miina commented on assets/src/edit-story/app/story/useStoryReducer/reducers/combineElements.js:
```diff
@@ -80,6 +80,7 @@ function combineElements(state, { firstId, firstElement, secondId }) {
     focalX = 50,
     focalY = 50,
     isFill = false,
+    flip = { vertical: false, horizontal: false },
```
Should we perhaps have a constant for these repeating default values?

barklund replied:
Done in 0994bb0 - though I'm not a huge fan of importing from elements here, I do think it's the best case. Alternatively we'd do this defaulting out where `combineElements` is called, but that's less nice.

barklund replied:
Found a better solution in 9dfc2d2e (also the other solution didn't work because of circular dependencies).

dvoytenko commented on assets/src/edit-story/components/dropTargets/provider.js:
```diff
@@ -82,11 +82,25 @@ function DropTargetsProvider({ children }) {
       }
 
+      const {
+        scale = 100,
```
Can we get these from `elements/x/index.js` `defaultAttributes`?

barklund replied:
Done in 0994bb0 though indirectly.

Commit:
Use element defaults

Commit:
Better location for default media attributes

Commit:
Merge branch 'master' into fix/1411-bg-consistency

miina reviewed (approved):
LGTM.

There's still the shapes + flipping issue but this might be a general issue instead: #1929

Commit:
Fix unit test

Commit:
Merge branch 'master' into fix/1411-bg-consistency

Commit:
[WIP] Added new test for bg drop target

Commit:
Fixed basic test case for bg droptarget

Commit:
Improved karma test with better flow and custom matchers

Commit:
Fixed unrelated eslint bug - maybe due to upgraded linter?

barklund commented on assets/src/edit-story/components/dropTargets/karma/background.karma.js:
```diff
@@ -0,0 +1,209 @@
+      // And verify that we no longer have a replacement element
+      const rep2 = await getCanvasBackgroundReplacement(fixture);
+      expect(rep2).toBeEmpty();
+    });
```
I feel this reads very well now, but it required quite a few helper functions and custom matchers to be this compact. This definitely has potential though - and runs quickly!

barklund commented on assets/src/edit-story/components/dropTargets/karma/background.karma.js:
```diff
@@ -0,0 +1,209 @@
+
+      // Verify that replacement img has correct source
+      const replaceImg = rep1.querySelector('img');
+      expect(replaceImg).toHaveSource(imageData.resource.src);
```
Yeah yeah, this matcher is of course way too specific, but just wanted to test the syntax out.

barklund replied:
A better matcher would be:

```js
      expect(replaceImg).toHaveProperty('src', imageData.resource.src);
```

dvoytenko commented on assets/src/edit-story/app/story/useStoryReducer/reducers/combineElements.js:
```diff
@@ -76,10 +81,11 @@ function combineElements(state, { firstId, firstElement, secondId }) {
+    focalX = DEFAULT_ATTRIBUTES_FOR_MEDIA.focalX,
+    focalY = DEFAULT_ATTRIBUTES_FOR_MEDIA.focalY,
+    isFill = DEFAULT_ATTRIBUTES_FOR_MEDIA.isFill,
+    flip = {},
```
What does empty `flip` mean?

barklund replied:
It's functionally the same as `flip: { vertical: false, horizontal: false }`, just simpler. Would you prefer it more explicit?

barklund replied:
Changed in dcc1ea85e (but essentially still the same).

dvoytenko replied:
Ok. I guess it's ok. Just wanted to clarify. Didn't realize we allowed an empty object for that.

dvoytenko commented on assets/src/edit-story/app/story/useStoryReducer/reducers/combineElements.js:
```diff
@@ -76,10 +81,11 @@ function combineElements(state, { firstId, firstElement, secondId }) {
-    focalX = 50,
-    focalY = 50,
-    isFill = false,
+    scale = DEFAULT_ATTRIBUTES_FOR_MEDIA.scale,
```
Why not just `...DEFAULT_ATTRIBUTES_FOR_MEDIA`?

barklund replied:
Hm, yes, this could be done cleaner...

barklund replied:
Changed in dcc1ea85e.

Commit:
Better custom matchers

Commit:
Added cleaner copying of properties in `combineElements`

Also fixed some things in `objectPick` util method and added basic unit tests for it.

Commit:
Added test for pressing "set as bg"

Commit:
Merge branch 'master' into fix/1411-bg-consistency

Commit:
Merge branch 'master' into fix/1411-bg-consistency

Commit:
Added a lot of drop target tests for bg replacement

Commit:
Merge branch 'master' into fix/1411-bg-consistency

# Conflicts:
#	assets/src/edit-story/app/story/useStoryReducer/reducers/combineElements.js
#	assets/src/edit-story/app/story/useStoryReducer/test/combineElements.js
#	assets/src/edit-story/components/canvas/displayElement.js
#	assets/src/edit-story/elements/media/index.js

Commit:
Hm, just manually click it then?
